File: tests/python/test_serialization.py
from __future__ import annotations

import os
import struct
from pathlib import Path

import numpy as np
import pytest
from vectorforge import FlatIndex


def test_roundtrip_tmp_path(tmp_path: Path) -> None:
    path = tmp_path / "flat.bin"
    db = np.array([[0.0, 0.0], [3.0, 4.0]], dtype=np.float32)
    index = FlatIndex(dim=2, metric="l2")
    index.add(db)
    index.save(os.fspath(path))

    loaded = FlatIndex(dim=1, metric="cosine")
    loaded.load(os.fspath(path))
    assert loaded.dim == 2
    assert loaded.metric == "l2"
    assert loaded.ntotal == 2

    q = np.array([[0.0, 0.0]], dtype=np.float32)
    ids_a, dist_a = index.search(q, k=2)
    ids_b, dist_b = loaded.search(q, k=2)
    np.testing.assert_array_equal(ids_a, ids_b)
    np.testing.assert_allclose(dist_a, dist_b, atol=0.0)


def test_load_rejects_invalid_metric(tmp_path: Path) -> None:
    path = tmp_path / "bad_metric.bin"
    path.write_bytes(b"VF01" + struct.pack("<IIIQ", 1, 2, 7, 0))
    index = FlatIndex(dim=2, metric="l2")
    with pytest.raises(RuntimeError):
        index.load(os.fspath(path))


def test_load_rejects_truncated_payload(tmp_path: Path) -> None:
    path = tmp_path / "truncated_payload.bin"
    payload = np.array([1.0, 2.0], dtype=np.float32).tobytes()
    path.write_bytes(b"VF01" + struct.pack("<IIIQ", 1, 2, 0, 2) + payload)
    index = FlatIndex(dim=2, metric="l2")
    with pytest.raises(RuntimeError):
        index.load(os.fspath(path))


def test_load_rejects_trailing_byte(tmp_path: Path) -> None:
    path = tmp_path / "trailing.bin"
    payload = np.array([1.0, 2.0], dtype=np.float32).tobytes()
    path.write_bytes(b"VF01" + struct.pack("<IIIQ", 1, 2, 0, 1) + payload + b"\x00")
    index = FlatIndex(dim=2, metric="l2")
    index.add(np.array([[0.0, 0.0]], dtype=np.float32))
    with pytest.raises(RuntimeError):
        index.load(os.fspath(path))
    assert index.dim == 2
    assert index.metric == "l2"
    assert index.ntotal == 1

File: tests/test_serialization.cpp
#include "vectorforge/flat_index.hpp"

#include <catch2/catch_test_macros.hpp>
#include <catch2/matchers/catch_matchers_floating_point.hpp>
#include <filesystem>
#include <fstream>
#include <string>
#include <vector>

using vectorforge::FlatIndex;

TEST_CASE("save then load preserves dim, metric, size, and search", "[serialization]") {
    const std::filesystem::path path =
        std::filesystem::temp_directory_path() / "vectorforge_flat_roundtrip.bin";

    FlatIndex index(2, "l2");
    const float data[] = {0.0f, 0.0f, 3.0f, 4.0f, 1.0f, 1.0f};
    index.add(data, 3);
    index.save(path.string());

    FlatIndex loaded(8, "cosine");
    loaded.load(path.string());
    REQUIRE(loaded.dim() == 2);
    REQUIRE(loaded.metric() == vectorforge::Metric::L2);
    REQUIRE(loaded.size() == 3);

    const float query[] = {0.0f, 0.0f};
    const auto before = index.search(query, 2);
    const auto after = loaded.search(query, 2);
    REQUIRE(after.ids == before.ids);
    REQUIRE_THAT(after.distances[0], Catch::Matchers::WithinAbs(before.distances[0], 0.0f));
    REQUIRE_THAT(after.distances[1], Catch::Matchers::WithinAbs(before.distances[1], 1e-6f));

    std::filesystem::remove(path);
}

TEST_CASE("load rejects a truncated file", "[serialization]") {
    const std::filesystem::path path =
        std::filesystem::temp_directory_path() / "vectorforge_truncated.bin";
    {
        std::ofstream out(path, std::ios::binary);
        out.write("VF01", 4);
    }
    FlatIndex index(2, "l2");
    REQUIRE_THROWS_AS(index.load(path.string()), std::runtime_error);
    std::filesystem::remove(path);
}
